### PR TITLE
Compute NLL loan scopes using the polonius model

### PR DESCRIPTION
**Pull request by lqd**
For a *location-insensitive* analysis (that is, without expressiveness improvements for users yet), this PR implements loans going out of scope using reachability and liveness, rather than checking if the issuing region's values contain a given CFG point. This is equivalent to NLL scopes and computes the same data.

r? @matthewjasper

A couple of notes:
- there are some assumptions about SCC representatives, placeholders, free regions, and member constraints that I believe hold, and they're documented in the code
- this passes all the UI tests with `-Zpolonius=next` -- the perf is [not terrible](https://github.com/rust-lang/rust/pull/112432#issuecomment-1749685862) and there are a bunch of ways to improve it in the future.
- there's a fixme left, hopefully Matthew you know a clean way to get the information it mentions.

### CODE REVIEW

**Comment by lqd** [comment]
@matthewjasper there are still things in there that I'd like to clean up, (e.g. I'm currently seeing whether some of the reachability computation can be done in liveness, and may try to make the look a bit more like the refactored CFG traversal) but I don't think they're absolutely necessary and won't change the core of the PR: you could already take a look whether you approve of the direction. It'd be greatly appreciated 🙏

**Review by matthewjasper — COMMENTED** [review]
Overall this approach seems fine as a first step.

It would be nice to have this "rebased" on the changes in #111759 to make determining the differences easier.

**Comment by lqd** [comment]
Thanks for taking a look.

> It would be nice to have this "rebased" on the changes in #111759 to make determining the differences easier.

Sure, I'll try to tackle that soon. It'll still have to be somewhat different, compared to diffing over https://github.com/lqd/rust/blob/eaddc3707520988f8bc2d267cb192d2f0c63ee80/compiler/rustc_borrowck/src/dataflow.rs#L155-L231 prior to the refactor, as #111759 and its followup make more use of regions as sets of points during the CFG walk.

**Comment by lqd** [comment]
let's see the hit liveness takes to do additional checks on whether polonius is enabled

@bors try @rust-timer queue

**Comment by lqd** [comment]
This also could be partly noise: cachegrind only shows a 0.1% diff on helloworld. And the bumps shown below from the last few days, are to similar icounts within 0.03% of this run.

image

That seems acceptable at least to start an initial review.

File: compiler/rustc_borrowck/src/dataflow.rs [hunk 1]
```diff
@@ -239,15 +240,196 @@ pub fn calculate_borrows_out_of_scope_at_location<'tcx>(
...
+            // The loan dies within the first block, we're done and can early return.
+            return;
+        }
+
+        // The loan is not dead. Add successor BBs to the work list, if necessary.
+        for succ_bb in first_bb_data.terminator().successors() {
+            if self.visited.insert(succ_bb) {
+                self.visit_stack.push(succ_bb);
+            }
+        }
```
Comment (jackh726) [inline compiler/rustc_borrowck/src/dataflow.rs:324]: I think you should be able to fold these into the loop below if you used a `loop` with `let mut block`.

Do you necessarily need to go from `first_lo` to `first_hi`? There couldn't be any kills above `first_lo` in the block? (I guess it's technically more work, but maybe not that relevant?)

Comment (jackh726) [reply compiler/rustc_borrowck/src/dataflow.rs:324]: Could still special case `first_lo` by moving that also into a `let mut first` out of the loop

File: compiler/rustc_borrowck/src/dataflow.rs [hunk 2]
```diff
@@ -239,15 +240,196 @@ pub fn calculate_borrows_out_of_scope_at_location<'tcx>(
...
+
+        // We may end up visiting `first_block` again. This is not an issue: we know at this point
+        // that the loan is not killed in the `first_lo..=first_hi` range, so checking the
+        // `0..first_lo` range and the `0..first_hi` range gives the same result.
+        while let Some(block) = self.visit_stack.pop() {
+            let bb_data = &self.body[block];
```
Comment (jackh726) [inline compiler/rustc_borrowck/src/dataflow.rs:330]: Question: is there a difference between `&self.body[block]` and `&self.body.basic_blocks[block]`?

File: compiler/rustc_borrowck/src/dataflow.rs [hunk 3]
```diff
@@ -239,15 +240,196 @@ pub fn calculate_borrows_out_of_scope_at_location<'tcx>(
...
+            }
+
+            // - the loan goes out of scope at `location` if it's not contained within any regions
+            // live at this point.
+            //
+            // FIXME: if the issuing region `i` can reach a live region `r` at point `p`, and `r` is
+            // live at point `q`, then it's guaranteed that `i` would reach `r` at point `q`.
+            // Reachability is location-insensitive, and we could take advantage of that, by jumping
+            // to a further point than just the next statement: we can jump to the furthest point
+            // within the block where `r` is live.
```
Comment (jackh726) [inline compiler/rustc_borrowck/src/dataflow.rs:385]: Could certainly do a binary search here - but I'm not sure the complexity would be helpful. What could be interesting is going from end to start (depends how often we expect loans to be live at the end of a block versus dying within (and where)

**Review by jackh726 — COMMENTED** [review]
I don't think there's anything here that would cause me not to feel comfortable r+ing, but some thoughts nonetheless :)

**Comment by jackh726** [comment]
@bors r+

I think we can get this into tree. It's still behind a flag, so not worried.

**Pull request merged**